### ISSUE TITLE
[Map] The "object as keys" case

### ISSUE THREAD

**Issue by zloirock** [DavidBruant/Map-Set.prototype.toJSON#1]
``` js
new Map([[{id: 1}, 42], [{id: 2}, 43]]).toJSON() // => '{"[object Object]":43}'

// way to fix 1, ugly:

Map.prototype.toJSON = function(){
  var o = Object.create(null);
  for(let [key, value] of this.entries()){
    o[JSON.stringify(key)] = value; // or another way to stringify key
  }
  return o;
}

new Map([[{id: 1}, 42], [{id: 2}, 43]]).toJSON() // => '{"{\"id\":1}":42,"{\"id\":2}":43}'

// way to fix 2, more simple and correct:

Map.prototype.toJSON = function(){
  return [...this];
}

new Map([[{id: 1}, 42], [{id: 2}, 43]]).toJSON() // => '[[{"id":1},42],[{"id":2},43]]'
```

P.S. Please, fix typo in readme:

``` javascript
for(let [key, value] of this.entries()){
  o[k] = v;
}
```

**Comment by DavidBruant** [comment]
(thanks for the typo, fixed)

The problem you're pointing is that in the proposal if a map has objects as keys, the serialization will be terrible. It is true, but it is as terrible than what you get if you do:

``` js
var o = {};
o[{id: 1}] = 42;
o[{id: 2}] = 43;
JSON.stringify(o);
```

The underlying issue is that the default `Object.prototype.toString` is terrible.
I have reserves on whether this change proposed to the language should make a bold statement on what a better serialization should be.
Indeed, it isn't that hard for others who want some serialization to do:

``` js
var tS = function(){ return JSON.stringify(this) }; // purposefully not arrow for dynamic this

new Map([[{id: 1, toString: tS}, 42], [{id: 2, toString: tS}, 43]]).toJSON()
```

Any other function can be picked. The toString can also be shared in the prototype.

I'm letting the issue open as it's important to discuss later down the road.

**Comment by zloirock** [comment]
@DavidBruant as I wrote above, stringify key is ugly way. More convenient to present map as array of entries, you can't simply do the same if map will be presented as object:

``` js
var map1 = new Map([[{id: 1}, 42], [{id: 2}, 43]]);

var json = JSON.stringify(map1); // => '[[{"id":1},42],[{"id":2},43]]'

var data = JSON.parse(json); // => [[{"id":1},42],[{"id":2},43]]

var map2 = new Map(data); // => Map {{"id":1}: 42, {"id":2}: 43}

for(let [{id}, value] of data){
  console.log(id);    // => 1, 2
  console.log(value); // => 42, 43
}
```

**Comment by ljharb** [comment]
I agree with https://github.com/DavidBruant/Map-Set.prototype.toJSON/issues/1#issuecomment-83794192 - I think an array of entries - ie, `Array.from(CreateMapIterator(this, "key+value")` as the sole body for `Map#toJSON` would make sense. It would avoid `__proto__` issues as well as allow for easy re-creation of a `Map` upon stringification.

**Comment by ljharb** [comment]
ping @DavidBruant What do you think about changing this to be entries instead of a null object?

**Comment by DavidBruant** [comment]
Sorry, I had missed https://github.com/DavidBruant/Map-Set.prototype.toJSON/issues/1#issuecomment-83794192

I was wrong when I mentioned a `bold statement` about a better serialization (I wasn't up-to-date the format used by `.entries` (CreateMapIterator w/ "key+value")).
The ability to re-initialize a map directly is extremely compelling.

I'll make the change. (keeping this issue open until I do)

**Comment by DavidBruant** [comment]
Closed via #5
